pkg/services/authn/grpcutils/inproc_exchanger.go

```diff
@@ -0,0 +1,87 @@
+package grpcutils
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/go-jose/go-jose/v3/jwt"
+	"github.com/grafana/authlib/authn"
+	"github.com/grafana/authlib/claims"
+	"github.com/grafana/grafana/pkg/infra/localcache"
+	"golang.org/x/sync/singleflight"
+)
+
+var (
+	timeNow = time.Now
+)
+
+const (
+	tokenLifetime       = 5 * time.Minute
+	inProcTokenCacheKey = "in-proc-access-token" // #nosec G101 not a hardcoded credential
+)
+
+type inProcExchanger struct {
+	cache   *localcache.CacheService
+	singlef singleflight.Group
+}
+
+func ProvideInProcExchanger() *inProcExchanger {
+	return &inProcExchanger{
+		cache:   localcache.New(5*time.Minute, 5*time.Minute),
+		singlef: singleflight.Group{},
+	}
+}
+
+func (e *inProcExchanger) Exchange(ctx context.Context, r authn.TokenExchangeRequest) (*authn.TokenExchangeResponse, error) {
```

Comment:
So we are not really building a jwt token here right? I wonder if we should just json ecode the claims and parse it on the other side so do we even need to "rotate" it here?